### PR TITLE
Implement a script to detect ABI changes.

### PR DESCRIPTION
**Pull request by coryan**
And add a Travis build to detect any ABI changes.  This is part of the fixes for #43.

Note that this does not stop the build when the ABI changes, that should be enabled
in a future PR, once we have convinced ourselves that the ABI is stable.  I want to
enable this now though, so we can troubleshoot any problems with the script.

### CODE REVIEW

File: ci/test-abi/google_cloud_cpp_common.expected.abi.dump [hunk 1]
```diff
@@ -0,0 +1,564 @@
...
+                            '16246' => {
+                                         'Data' => 1,
+                                         'Header' => 'build_info.h',
+                                         'Line' => '25',
+                                         'MnglName' => '_ZN6google5cloud2v08internal8COMPILERE',
+                                         'NameSpace' => 'google::cloud::v0::internal',
+                                         'Return' => '16316',
+                                         'ShortName' => 'COMPILER',
+                                         'Source' => 'build_info.cc',
+                                         'SourceLine' => '22'
```
Comment (mbrukman) [inline ci/test-abi/google_cloud_cpp_common.expected.abi.dump]: Almost certainly line numbers will change when source code changes...

File: ci/test-abi/google_cloud_cpp_common.expected.abi.dump [hunk 2]
```diff
@@ -0,0 +1,564 @@
...
+          'Sources' => {
+                         'build_info.cc' => 1,
+                         'setenv.cc' => 1,
+                         'throw_delegate.cc' => 1
+                       },
+          'SymbolInfo' => {
+                            '16246' => {
+                                         'Data' => 1,
+                                         'Header' => 'build_info.h',
+                                         'Line' => '25',
```
Comment (mbrukman) [inline ci/test-abi/google_cloud_cpp_common.expected.abi.dump]: This won't change?

File: ci/test-abi/google_cloud_cpp_common.expected.abi.dump [hunk 3]
```diff
@@ -0,0 +1,564 @@
...
+                        'libstdc++.so.6' => 1
+                      },
+          'PublicABI' => '1',
+          'Sources' => {
+                         'build_info.cc' => 1,
+                         'setenv.cc' => 1,
+                         'throw_delegate.cc' => 1
+                       },
+          'SymbolInfo' => {
+                            '16246' => {
```
Comment (mbrukman) [inline ci/test-abi/google_cloud_cpp_common.expected.abi.dump]: Do these numbers really stay static across builds, compilers, etc.? Or are they internal numbers, and it is just a marker, but is compatible as long as there's correspondence between the id matching for function input / output args...?

File: ci/test-abi/google_cloud_cpp_common.expected.abi.dump (hunk 1 above)
Comment (coryan) [reply ci/test-abi/google_cloud_cpp_common.expected.abi.dump]: Yes.  The line numbers would change, but that does not affect the compatibility report.  For example, I made a local change to remove a function, which produced 
[compat_report.html.gz](https://github.com/GoogleCloudPlatform/google-cloud-cpp/files/2078024/compat_report.html.gz).  The line number changes had no effect, but the removed function is clearly identified.

I expect that we will commit new versions of this file only when the reports fail, which would  create a massive diff, but who cares?

File: ci/test-abi/google_cloud_cpp_common.expected.abi.dump (hunk 2 above)
Comment (coryan) [reply ci/test-abi/google_cloud_cpp_common.expected.abi.dump]: See above.

File: ci/test-abi/google_cloud_cpp_common.expected.abi.dump (hunk 3 above)
Comment (coryan) [reply ci/test-abi/google_cloud_cpp_common.expected.abi.dump]: These numbers also can change from build to build, but the compatibility reports are not affected.

**Comment by coryan** [comment]
Answered questions and rebased. PTAL.

File: ci/check-abi.sh [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+
+set -eu
+
+readonly BINDIR="$(dirname $0)"
+source "${BINDIR}/colors.sh"
+readonly PROJECT_ROOT="$(cd ${BINDIR}/..; pwd)"
+
+if [ "${CHECK_ABI:-}" != "yes" ]; then
+  echo
+  echo "${COLOR_YELLOW}Skipping abi check as it is disabled for this build." \
```
[the remark this answers is not in the document]
Comment (coryan) [reply ci/check-abi.sh]: Fixed.

**Pull request merged**